src/raglite/_extract.py

```diff
@@ -88,3 +97,275 @@ class MyNameResponse(BaseModel):
+    key = field_spec["key"]
+    field_type = field_spec["type"]
+    prompt = field_spec["prompt"]
+    source = field_spec.get("source", "content")
+    max_chars = field_spec.get("max_chars")
+
+    values, is_multi_value = _extract_literal_values(field_type)
+
+    if source == "content":
+        source_text = doc.content
+        source_label = "document content"
+        if source_text is None or source_text.strip() == "":
+            msg = "Document content is empty; cannot extract metadata from content."
+            raise ValueError(msg)
+    else:
+        source_text = doc.metadata_.get(source)
+        source_label = f"metadata field '{source}'"
+        if source_text is None:
+            msg = f"Metadata field '{source}' not found in document metadata."
+            raise ValueError(msg)
+
+    # Apply character limit if specified
+    if max_chars is not None and isinstance(source_text, str) and len(source_text) > max_chars:
+        source_text = source_text[:max_chars]
+        source_label += f" (truncated to {max_chars} chars)"
+
+    field_prompt = f"**{key}**: {prompt}\n"
+    field_prompt += f"  - Source: {source_label}\n"
+    field_prompt += f"  - Text to analyze: {source_text}\n"
+
+    # TODO: Determine if we want to add this also when supporting response_format
+    if values:  # Literal fields (single or multi-value)
+        values_str = ", ".join(f'"{val}"' for val in values)
+        if is_multi_value:
+            field_prompt += f"  - Allowed values: {values_str}\n"
+            field_prompt += "  - Select ALL that apply\n"
+        else:
+            field_prompt += f"  - Allowed values: {values_str}\n"
+            field_prompt += "  - Select ONE that best fits\n"
+    elif is_multi_value:  # Multi-value free-text field
+        field_prompt += "  - Provide relevant values (can be multiple)\n"
+    else:  # Single-value free-text field
+        field_prompt += "  - Provide a single, concise response\n"
+
+    return field_prompt + "\n"
+
+
+def expand_document_metadata(  # noqa: C901, PLR0912, PLR0915
+    documents: list[Document],
+    metadata_fields: list[dict[str, Any]],
+    config: RAGLiteConfig,
+    *,
+    max_concurrent_requests: int | None = None,
+) -> Iterator[Document]:
+    """
+    Extract metadata from documents using LLM with configurable fields and constraints.
+
+    Parameters
+    ----------
+    documents : list[Document]
+        List of documents to process
+    metadata_fields : list[dict[str, Any]]
+        List of metadata field specifications. Each dict should contain:
+        - key (str): The metadata field name
+        - type (Any): The exact type for the field:
+            (a.) Literal["A","B"] for single-choice,
+            (b.) list[Literal["X","Y","Z"]] for multi-choice,
+            (c.) str, int, float, bool for single-value free-text fields, and
+            (d.) list[str], list[int], list[float], list[bool] for multi-value free-text fields
```

Comment:
Should we support these two from the beginning? I would prefer to be defensive in terms of what we offer here until we're sure we can support this type in the database filter itself.